### PR TITLE
More collect_column fixes

### PR DESCRIPTION
**Pull request by piever**
Correct function to choose whether to widen correctly in all cases, correctly create `DataValueArray` even if column starts as regular `Array` but widens later.

### CODE REVIEW

File: src/collect.jl [hunk 1]
```diff
@@ -1,8 +1,14 @@
...
-_is_subtype(::Type{DataValue{S}}, ::Type{DataValue{T}}) where {S, T} = S<:T
+_is_subtype(::Type{S}, ::Type{T}) where {S, T} = promote_type(S, T) == T
+
+Base.@pure function dataarrayof(T)
+    if T<:DataValue
+        DataValueArray{T.parameters[1],1}
+    else
+        Vector{T}
+    end
+end
```
Comment (shashi) [inline src/collect.jl]: There's already an `arrayof` function....

Comment (piever) [reply src/collect.jl]: But that one doesn't quite do it because it would create a `Columns` object when starting from a `Tuple`, which I don't think we want. For example `[((1,2), (2,3))]` should in my view give `Columns([(1,2)], [(2,3)])` rather than a nested columns object

Comment (shashi) [reply src/collect.jl]: Wouldn't Columns of Columns be more efficient storage? (e.g. in the case where the tuple contains strings.)

What makes it undesirable? It should have the same properties as a Vector of tuples.

Comment (piever) [reply src/collect.jl]: The more I think about it the more I agree with you, nested tuples should nest, `dataarrayof` can be replaced by `arrayof`, `fieldwise_isa` and `widencolumns` should be recursive and I'd need to add some tests for the nested corner cases.
